### PR TITLE
Python Solution for Exercise Evaluation 1,2,3

### PR DESCRIPTION
**Pull request by StefanieSwz**
General:
Added: 
-exercises/evaluation/I2ML_env.yml
-exercises/data/german_credit_for_py.csv

Evaluation 1:
Modified:
- exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw
- exercises/evaluation/ex_rnw/ex_train_test_split.Rnw
Added:
- exercises/evaluation/sol_eval_1_py.ipynb

Evaluation 2:
Modified:
- exercises/evaluation/ex_rnw/ex_resa_mlr3.Rnw
Added: 
- exercises/evaluation/sol_eval_2_py.ipynb

Evaluation 3:
Modified: 
- exercises/evaluation/ex_rnw/ex_roc_2.Rnw
Added:
- exercises/evaluation/sol_eval_3_py.ipynb

Please note: color coding for python code snippts not working currently

### CODE REVIEW

**Comment by StefanieSwz** [comment]
[Python Exercise I2ML: Evaluation 1,2,3](https://trello.com/c/ZR60HMlx/285-python-exercise-i2ml-evaluation-123)

File: exercises/evaluation/sol_eval_1_py.ipynb
Comment (ludwigbothmann) [inline exercises/evaluation/sol_eval_1_py.ipynb:1]: 1)
- Was macht das `poly.fit()` - poly wird danach ja nicht mehr benutzt und die Transformation wurde ja schon mit `poly.fit_transform` gemacht?
- Die Ergebnisse sind nicht identisch wie in der R-Lösung, aber das ist vmtl. auch nicht machbar wegen der simulierten Daten, oder? (Also außer wenn man die Daten von R nimmt, aber das finde ich blöd - bottomline ist ja die gleich
- fixed some typos
2) Gibt es einen speziellen Grund, dass die x-Achse anders sortiert ist als in der R-Lösung? Wenn nicht, dann dreh es vielleicht besser um, dann sieht man gleich, dass es passt
Sonst sehr schön, mal wieder, danke :)

File: exercises/evaluation/sol_eval_2_py.ipynb
Comment (ludwigbothmann) [inline exercises/evaluation/sol_eval_2_py.ipynb:1]: 2)
- b) weißt Du, wieso da nicht das gleiche rauskommt wie in R? (bei c liegt es ja dann am random seed vermute ich)
sonst passt es denke ich

File: exercises/evaluation/ex_rnw/figure/unnamed-chunk-1-1.pdf
Comment (ludwigbothmann) [inline exercises/evaluation/ex_rnw/figure/unnamed-chunk-1-1.pdf:1]: Das brauchen wir denke ich nicht im Repo, ist ja nur eine temporäre Datei beim kompilieren, oder?

File: exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw
Comment (ludwigbothmann) [inline exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw:1]: Hier und auch bei der anderen Aufgabe: Der Python-Code scheint nun doch ausgewertet zu werden, was bei mir zu einem Haufen Fehler führt, zB weil er `pandas` nicht findet (obwohl ich es aus dem richtigen env heraus kompiliere). Könntest Du das noch so einstellen, dass der Python-Code zwar angezeigt, aber nicht ausgeführt wird?

File: exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw [hunk 1]
```diff
@@ -56,6 +55,25 @@ y_test <- 10 + 3 * sin(0.15 * pi * x_test) + rnorm(length(x_test), sd = 0.5)
...
+<<echo=TRUE, message=FALSE, warning=FALSE, engine="python">>=
+import numpy as np
+import pandas as pd
+import math
+
+np.random.seed(43)
+x_train = np.linspace(10, 15, num = 50)
+y_train = 10 + 3 * np.sin(0.15 * math.pi * x_train) + \
+      np.random.normal(loc=0.0, scale=0.5, size= len(x_train))
+data_train = pd.DataFrame({'y': y_train, 'x': x_train})
```
[the remark this answers is not in the document]
Comment (StefanieSwz) [reply exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw]: Gut, ist ausgetauscht!

File: exercises/evaluation/sol_eval_1_py.ipynb
Comment (StefanieSwz) [reply exercises/evaluation/sol_eval_1_py.ipynb:1]: 1.
- Ja stimmt, das poly.fit() ist redundant, ist jetzt gelöscht!
- Ja genau, ich habe hier echt ein bisschen rumprobiert bis der seed gepasst hat, um den Fall aus der Übung nachstellen zu können. Dadurch, dass man in der Angabe den Seed gegeben bekommt sollte es vom inhaltlichen passen mMn.
2. Ok, auch drinnen :)

File: exercises/evaluation/sol_eval_2_py.ipynb
Comment (StefanieSwz) [reply exercises/evaluation/sol_eval_2_py.ipynb:1]: Ja, das ist spannend. Die Daten habe ich wieder aus der mlr3 task gezogen, doch der cleaning Aufwand ist bei 20 features relativ hoch, deswegen hier das präparierte csv file. Die ordinalen features werden in Python trotzdem noch anders verarbeitet werden als in R. Man induziert ja in R die Ordnung auf kategorielle Variablen und entsprechend weiter verwendet, während in sklearn numerische features verlangt werden. Nach dem Ordinal Encoding bekommt man also z.B. den Vektor [0,2,0,1,1,2,...] und dieser wird normal gefitted. Hier also ein Unterschied.
Für die anderen kategoriellen Variablen ist jetzt auch nochmal ein Update dabei, dass pro feature eine dummy Variable gedropt wird.
Insgesamt komme ich so auf einen MCE von 0.2149, was sich zumindest auf die Lösung in mlr3 runden lässt

File: exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw
Comment (StefanieSwz) [reply exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw:1]: Ja, ist drinn!

File: exercises/evaluation/ex_rnw/figure/unnamed-chunk-1-1.pdf
Comment (StefanieSwz) [reply exercises/evaluation/ex_rnw/figure/unnamed-chunk-1-1.pdf:1]: Ja genau, das war schon im letzten commit, soweit ich weiß

File: exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw
Comment (ludwigbothmann) [reply exercises/evaluation/ex_rnw/ex_regr_metrics.Rnw:1]: Top, jetzt läuft es durch :+1:

**Review by ludwigbothmann — APPROVED** [review]
Super, danke für die Änderungen!
